Atomic.md

```diff
@@ -0,0 +1,278 @@
+Atomic placeholder operations in VFSForGit on MacOS
+===================================================
+
+A potential hole in our current design is in our creation of
+placeholders. Today, when creating a placeholder directory, at a high
+level, we go through the following workflow:
+
+-   Create placeholder folder/file in the virtualization root
+
+-   'Mark up' the file with the proper flags/xattrs
+
+-   (If creating a file) Update the mode of the file
+
+This could result in scenarios where a incomplete placeholder appears in
+the virtualization root. The full set of failure modes is documented
+below:
+
+WritePlaceholderDirectory:
+
+-   mkdir() fails to create the placeholder directory
+
+    -   The enumeration fails as a directory wasn't created. This should
+         be retriable as a second attempt should see the IsEmpty bit on
+         the parent folder and call WritePlaceholderDirectory again. No
+         incomplete placeholders were left in the parent directory.
```

Comment:
You're assuming here that mkdir failed on the creation of the first placeholder within the parent directory. If this is the second or subsequent placeholder, then the parent now has children in it, and the next attempt to enumerate the parent will fail.

Reply:
Ah, I wrote this section with the design that is proposed below in mind. I'll rework that.